fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarCFRPresenter.kt

```diff
@@ -51,56 +56,80 @@ private const val CRF_MINIMUM_NUMBER_OPENED_TABS = 5
  * @param settings used to read and write persistent user settings
  * @param toolbar will serve as anchor for the CFRs
  * @param sessionId optional custom tab id used to identify the custom tab in which to show a CFR.
+ * @param shoppingExperienceFeature optional reference to [ShoppingExperienceFeature].
  */
 class BrowserToolbarCFRPresenter(
     private val context: Context,
     private val browserStore: BrowserStore,
     private val settings: Settings,
     private val toolbar: BrowserToolbar,
     private val sessionId: String? = null,
+    private val shoppingExperienceFeature: ShoppingExperienceFeature = ShoppingExperienceFeature(),
 ) {
     @VisibleForTesting
-    internal var tcpCfrScope: CoroutineScope? = null
+    internal var scope: CoroutineScope? = null
 
     @VisibleForTesting
-    internal var tcpCfrPopup: CFRPopup? = null
+    internal var popup: CFRPopup? = null
 
     /**
      * Start observing [browserStore] for updates which may trigger showing a CFR.
      */
     @Suppress("MagicNumber")
     fun start() {
-        if (shouldShowCFR()) {
-            tcpCfrScope = browserStore.flowScoped { flow ->
-                flow
-                    .mapNotNull { it.findCustomTabOrSelectedTab(sessionId)?.content?.progress }
-                    // The "transformWhile" below ensures that the 100% progress is only collected once.
-                    .transformWhile { progress ->
-                        emit(progress)
-                        progress != 100
-                    }
-                    .filter { it == 100 }
-                    .collect {
-                        tcpCfrScope?.cancel()
-                        showTcpCfr()
-                    }
+        when (getCFRToShow()) {
+            ToolbarCFR.TCP -> {
+                scope = browserStore.flowScoped { flow ->
+                    flow.mapNotNull { it.findCustomTabOrSelectedTab(sessionId)?.content?.progress }
+                        // The "transformWhile" below ensures that the 100% progress is only collected once.
+                        .transformWhile { progress ->
+                            emit(progress)
+                            progress != 100
+                        }.filter { it == 100 }.collect {
+                            scope?.cancel()
+                            showTcpCfr()
+                        }
+                }
+            }
+
+            ToolbarCFR.SHOPPING -> {
+                scope = browserStore.flowScoped { flow ->
+                    flow.mapNotNull { it.selectedTab }
+                        .filter { it.isProductUrl && it.content.progress == 100 && !it.content.loading }
+                        .distinctUntilChanged()
+                        .collect {
+                            if (toolbar.findViewById<View>(R.id.mozac_browser_toolbar_page_actions).isVisible) {
+                                scope?.cancel()
+                                showShoppingCFR()
+                            }
```

Comment:
should this cancel the scope regardless of the `if`? (asking since I don't know the ins and outs of this feature)

Reply:
During my testing I found that sometimes the current tab may load a product page but no shopping cart would be shown. My idea was the CFR would then be loaded for the next product page during the same browsing session. This way the CFR would present the shopping cart icon on its first appearance.